protocols/floodsub/src/layer.rs

```diff
@@ -0,0 +1,277 @@
+        for (peer_id, sub_topic) in self.connected_peers.iter() {
+            if !sub_topic.iter().any(|t| message.topics.iter().any(|u| t == u)) {
+                continue;
+            }
+
+            self.events.push_back(NetworkBehaviorAction::SendEvent {
+                peer_id: peer_id.clone(),
+                event: FloodsubRpc {
+                    subscriptions: Vec::new(),
+                    messages: vec![message.clone()],
+                }
+            });
+        }
+    }
+
+    /// Builds a unique sequence number to put in a `FloodsubMessage`.
+    fn next_sequence_number(&mut self) -> Vec<u8> {
+        let data = self.seq_no.to_string();
+        self.seq_no += 1;
+        data.into()
+    }
+}
+
+impl<TSubstream> NetworkBehavior for FloodsubBehaviour<TSubstream>
+where
+    TSubstream: AsyncRead + AsyncWrite + Send + Sync + 'static,
+{
+    type ProtocolsHandler = FloodsubHandler<TSubstream>;
+    type OutEvent = FloodsubMessage;
+
+    fn new_handler(&mut self) -> Self::ProtocolsHandler {
+        FloodsubHandler::new()
+    }
+
+    fn inject_connected(&mut self, id: PeerId, _: ConnectedPoint) {
+        // We need to send our subscriptions to the newly-connected node.
+        for topic in self.subscribed_topics.iter() {
+            self.events.push_back(NetworkBehaviorAction::SendEvent {
+                peer_id: id.clone(),
+                event: FloodsubRpc {
+                    messages: Vec::new(),
+                    subscriptions: vec![FloodsubSubscription {
+                        topic: topic.hash().clone(),
+                        action: FloodsubSubscriptionAction::Subscribe,
+                    }],
+                },
+            });
+        }
+
+        self.connected_peers.insert(id.clone(), SmallVec::new());
+    }
+
+    fn inject_disconnected(&mut self, id: &PeerId, _: ConnectedPoint) {
+        let was_in = self.connected_peers.remove(id);
+        debug_assert!(was_in.is_some());
+    }
+
+    fn inject_node_event(
+        &mut self,
+        propagation_source: PeerId,
+        event: FloodsubRpc,
+    ) {
+        // List of messages we're going to propagate on the network.
+        let mut rpcs_to_dispatch: Vec<(PeerId, FloodsubRpc)> = Vec::new();
+
+        for message in event.messages {
+            // Use `self.received` to skip the messages that we have already received in the past.
+            // Note that this can false positive.
+            if !self.received.test_and_add(&message) {
+                continue;
+            }
+
+            // Add the message to be dispatched to the user.
+            if self.subscribed_topics.iter().any(|t| message.topics.iter().any(|u| t.hash() == u)) {
+                self.events.push_back(NetworkBehaviorAction::GenerateEvent(message.clone()));
+            }
+
+            // Propagate the message to everyone else who is subscribed to any of the topics.
+            for (peer_id, sub_topics) in self.connected_peers.iter() {
```

Comment:
`sub_topics` sounds a little like `sub topic`, i.e. like there's a hierarchy of topics. Maybe change to just `topics` to avoid that?